### DISCUSSION TITLE
Help with syntax for nested combos and macros.

### DISCUSSION THREAD

**Discussion by BornOfStygian** [sezanzeb/input-remapper#166] [Q&A]
I recently found key-mapper and I am loving the user friendly gui and flexibility.
The documentation is great and  the cross device example seems like it is very close to what I want. I have a standard 5 button mouse and what I'm trying to do is have the buttons on the side (SIDE,EXTRA) function normally if pressed and released without any other mapped combination, but work as their own modifier in combination with other buttons on the same device. For some reason the syntax just isn't clicking in my brain. It would look something like this without exact naming:

SIDE = released alone = SIDE
+LEFT = META+LEFT
+RIGHT = META+RIGHT
+SCROLL_UP = L_CTRL+EQUAL
+SCROLL_DOWN = L__CTRL+MINUS
EXTRA = released alone = EXTRA
+LEFT = r(LEFT) *auto-clicker*
+RIGHT = r(RIGHT) *auto-clicker*

I usually remap games for ESDF instead of WASD to maximize available keys around primary controls. So I already remapped capslock which is premo real-estate and finally have my double-sided return keyboard.  Now I want to try doing more in-depth stuff and maybe function nesting if possible. 

If anyone has any recommended exercises to practice or advice on what I'd like done I would appreciate it. Usually when example are provided like in the material I don't have any problem filling in the blanks. Just trying to wrap my head around it. Any help is appreciated.

**Comment by cybermaus** [comment]
Without testing/trying, some suggestions:

I think SCROLL_UP/DOWN cannot be used as input. They are EV_REL events and not EV_KEY or EV_BTN events, and if I am correct this tool can send some simple EV_REL events, but cannot use them as input. But such a fancy mouse also scroll-left/right buttons probably?

Edit: turn out SCROLL_UP/DOWN is available as input event. It is the only (at the moment) EV_REL event that can be used, so the general suggestion you cannot listen to EV_REL events is correct, but scroll has been implemented.

As to using SIDE as modifier key
```
BTN_SIDE => set(side, 1).h(BTN_SIDE).set(side, 0) 
BTN_LEFT => ifeq(side,1,m(KEY_META,k(BTN_LEFT))
BTN_RIGHT => ifeq(side,1,m(KEY_META,k(BTN_RIGHT))
```

It is possible to make also the side button by itself work, you need to add some juggling
```
BTN_SIDE => set(side, 1).h(BTN_SIDE).ifeq(side,1,k(BTN_SIDE)).set(side, 0) 
BTN_LEFT => ifeq(side,1,m(KEY_META,k(BTN_LEFT).set(side,0))
BTN_RIGHT => ifeq(side,1,m(KEY_META,k(BTN_RIGHT).set(side,0))
```

That has as a result you cannot use both left and right while holding side down. If you want that too, I think you need to split the variables.<br/>May as well start using proper variable names, it's no longer going to be small anyway.
```
BTN_SIDE => set(sidedown, 1).set(sideunused, 1).h(BTN_SIDE).ifeq(sideunused,1,k(BTN_SIDE)).set(sidedown, 0).set(sideunused, 0)
BTN_LEFT => ifeq(sidedown,1,m(KEY_META,k(BTN_LEFT).set(sideunused,0))
BTN_RIGHT => ifeq(sidedown,1,m(KEY_META,k(BTN_RIGHT).set(sideunused,0))
```

**Reply by BornOfStygian** [reply]
Thanks, that next to the key-mapper/readme/macros.md page really helps this make more sense. I realize now that this maybe because I haven't slept in over 24 hours, sometime after the dozenth attempt to read your reply. It already helps demystify what's going on. I'll take another crack at it after some rest. Just didn't want to leave you hanging and say thank you.

**Reply by BornOfStygian** [reply]
Doing some testing after a nap, wheel(up, 1).wheel(down, 1).wheel(left, 1).wheel(right, 1) using the autofill from the mapping field all do what I need them to do in this particular case. Just though I would let you know in case it ever comes up again. Thanks again for the assist.

**Reply by sezanzeb** [reply]
thanks @cybermaus for helping. So is everything clear now?

**Reply by BornOfStygian** [reply]
It's "clearer"
Appreciate the quick reply. I now have a seconds mouse pluged in for testing as I get used to to the macros and limitations. Got stuff functional but I'm going through the rounds to make sure I understand it instead of just coping an example someone was nice enough to provide. This is the perfect example of software that I like to recommend, I'm trying to learn Autokey as well but key-mapper is far more practical for 99% of people I know would need something like this.  Just trying to make sure I have a grip on it before recommending it so if they need any help I can chime in like cybermaus. Context profiles with holding a button was the first thing I thought of because I use it and nested menus in steam input all the time.
